src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,78 @@
+        save(new Product("xperiaxz", "Sony Xperia XZ", new BigDecimal(120), usd, 100, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Sony/Sony%20Xperia%20XZ.jpg"));
+        save(new Product("nokia3310", "Nokia 3310", new BigDecimal(70), usd, 100, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Nokia/Nokia%203310.jpg"));
+        save(new Product("palmp", "Palm Pixi", new BigDecimal(170), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Palm/Palm%20Pixi.jpg"));
+        save(new Product("simc56", "Siemens C56", new BigDecimal(70), usd, 20, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C56.jpg"));
+        save(new Product("simc61", "Siemens C61", new BigDecimal(80), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C61.jpg"));
+        save(new Product("simsxg75", "Siemens SXG75", new BigDecimal(150), usd, 40, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20SXG75.jpg"));
+    }
+
     @Override
-    public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+    public Product getProduct(Long id) throws NoProductWithSuchIdException {
+        synchronized (mutex) {
+            return products.stream()
+                    .filter(product -> product.getId().equals(id))
+                    .findAny()
+                    .orElseThrow(() -> new NoProductWithSuchIdException(id));
+        }
     }
 
     @Override
     public List<Product> findProducts() {
-        throw new RuntimeException("Not implemented");
+        return products.stream()
+                .filter(product -> product.getPrice() != null)
+                .filter(product -> product.getStock() > 0)
+                .collect(Collectors.toList());
     }
 
     @Override
     public void save(Product product) {
-        throw new RuntimeException("Not implemented");
+        synchronized (mutex) {
+            product.setId(maxId++);
+            products.add(product);
+        }
     }
 
     @Override
-    public void delete(Long id) {
-        throw new RuntimeException("Not implemented");
+    public void delete(Long id) throws NoProductWithSuchIdException {
+        // Create key
+        Product key = new Product();
+        key.setId(id);
+        synchronized (mutex) {
+            // Find index
+            int index = Collections.binarySearch(products, key, Comparator.comparing(Product::getId));
+            // Check if item is found
+            if (index >= 0) {
+                products.remove(index);
+            } else {
+                throw new NoProductWithSuchIdException(id);
```

Comment:
Can we use getProduct(id) here instead of binarySearch?

Or even something like this:

products.removeIf(product -> id.equals(product.getId()));

Reply:
We couldn't use getProduct as we needed list index of product with a specified ID to remove this product by that index.
We could use removeIf, but time complexity of removeIf is O(n) and total time complexity of binary search and remove by index is O(log n).
Now see new implementation with Map instead of List.